File: patterns/architectural/event-driven.py
"""
Event-Driven Architecture Pattern - Python

Architecture where components communicate through events.
Decouples producers and consumers of events.
"""


# Event Bus
class EventBus:
    def __init__(self):
        self.listeners = {}
    
    def on(self, event, callback):
        if event not in self.listeners:
            self.listeners[event] = []
        self.listeners[event].append(callback)
    
    def off(self, event, callback):
        if event in self.listeners:
            self.listeners[event] = [cb for cb in self.listeners[event] if cb != callback]
    
    def emit(self, event, data):
        if event in self.listeners:
            for callback in self.listeners[event]:
                callback(data)


# Event Producers
class OrderService:
    def __init__(self, event_bus):
        self.event_bus = event_bus
        self.orders = []
    
    def create_order(self, order_data):
        import time
        order = {
            'id': int(time.time()),
            **order_data,
            'status': 'created'
        }
        self.orders.append(order)
        
        # Emit event
        self.event_bus.emit('order.created', order)
        return order
    
    def complete_order(self, order_id):
        order = next((o for o in self.orders if o['id'] == order_id), None)
        if order:
            order['status'] = 'completed'
            self.event_bus.emit('order.completed', order)


# Event Consumers
class EmailService:
    def __init__(self, event_bus):
        self.event_bus = event_bus
        self.setup_listeners()
    
    def setup_listeners(self):
        self.event_bus.on('order.created', self.send_order_confirmation)
        self.event_bus.on('order.completed', self.send_order_completion)
    
    def send_order_confirmation(self, order):
        print(f"Email: Order {order['id']} confirmation sent to {order['customer_email']}")
    
    def send_order_completion(self, order):
        print(f"Email: Order {order['id']} completion notification sent to {order['customer_email']}")


class InventoryService:
    def __init__(self, event_bus):
        self.event_bus = event_bus
        self.inventory = {'item1': 100, 'item2': 50}
        self.setup_listeners()
    
    def setup_listeners(self):
        self.event_bus.on('order.created', self.reserve_items)
    
    def reserve_items(self, order):
        for item in order['items']:
            if item['id'] in self.inventory:
                self.inventory[item['id']] -= item['quantity']
                print(f"Inventory: Reserved {item['quantity']} of {item['id']}")


# Example usage
if __name__ == "__main__":
    print("=== Event-Driven Architecture Pattern ===\n")
    
    event_bus = EventBus()
    order_service = OrderService(event_bus)
    email_service = EmailService(event_bus)
    inventory_service = InventoryService(event_bus)
    
    order = order_service.create_order({
        'customer_email': 'customer@example.com',
        'items': [
            {'id': 'item1', 'quantity': 2},
            {'id': 'item2', 'quantity': 1}
        ]
    })
    
    print()
    order_service.complete_order(order['id'])

File: patterns/architectural/event-driven.js
/**
 * Event-Driven Architecture Pattern - JavaScript
 * 
 * Architecture where components communicate through events.
 * Decouples producers and consumers of events.
 */

// Event Bus
class EventBus {
    constructor() {
        this.listeners = {};
    }
    
    on(event, callback) {
        if (!this.listeners[event]) {
            this.listeners[event] = [];
        }
        this.listeners[event].push(callback);
    }
    
    off(event, callback) {
        if (this.listeners[event]) {
            this.listeners[event] = this.listeners[event].filter(cb => cb !== callback);
        }
    }
    
    emit(event, data) {
        if (this.listeners[event]) {
            this.listeners[event].forEach(callback => callback(data));
        }
    }
}

// Event Producers
class OrderService {
    constructor(eventBus) {
        this.eventBus = eventBus;
        this.orders = [];
    }
    
    createOrder(orderData) {
        const order = {
            id: Date.now(),
            ...orderData,
            status: 'created'
        };
        this.orders.push(order);
        
        // Emit event
        this.eventBus.emit('order.created', order);
        return order;
    }
    
    completeOrder(orderId) {
        const order = this.orders.find(o => o.id === orderId);
        if (order) {
            order.status = 'completed';
            this.eventBus.emit('order.completed', order);
        }
    }
}

// Event Consumers
class EmailService {
    constructor(eventBus) {
        this.eventBus = eventBus;
        this.setupListeners();
    }
    
    setupListeners() {
        this.eventBus.on('order.created', (order) => {
            this.sendOrderConfirmation(order);
        });
        
        this.eventBus.on('order.completed', (order) => {
            this.sendOrderCompletion(order);
        });
    }
    
    sendOrderConfirmation(order) {
        console.log(`Email: Order ${order.id} confirmation sent to ${order.customerEmail}`);
    }
    
    sendOrderCompletion(order) {
        console.log(`Email: Order ${order.id} completion notification sent to ${order.customerEmail}`);
    }
}

class InventoryService {
    constructor(eventBus) {
        this.eventBus = eventBus;
        this.inventory = { 'item1': 100, 'item2': 50 };
        this.setupListeners();
    }
    
    setupListeners() {
        this.eventBus.on('order.created', (order) => {
            this.reserveItems(order.items);
        });
    }
    
    reserveItems(items) {
        items.forEach(item => {
            if (this.inventory[item.id]) {
                this.inventory[item.id] -= item.quantity;
                console.log(`Inventory: Reserved ${item.quantity} of ${item.id}`);
            }
        });
    }
}

// Example usage
console.log('=== Event-Driven Architecture Pattern ===\n');

const eventBus = new EventBus();
const orderService = new OrderService(eventBus);
const emailService = new EmailService(eventBus);
const inventoryService = new InventoryService(eventBus);

const order = orderService.createOrder({
    customerEmail: 'customer@example.com',
    items: [
        { id: 'item1', quantity: 2 },
        { id: 'item2', quantity: 1 }
    ]
});

console.log();
orderService.completeOrder(order.id);

module.exports = { EventBus, OrderService, EmailService, InventoryService };

